src/frontend/src/helpers/Storage/WebStorage.ts

```diff
@@ -0,0 +1,588 @@
+  }
+
+  public async invoke_cb(type: string, message?: any): Promise<Array<any>> {
+    return this.callbacks.filter(
+      (x: Callback) => { return x && x.type === type; }).map(
+        async (x: Callback) => { return x.cb(message); });
+  }
+
+  // wrapper function for invoke_cb meant to be used to call
+  //  the websocket failure coniditons 'disconnected' and 'failure'
+  public invoke_cb_failure_om_wrap(type: string, message?: any): PromiseLike<Array<any>> {
+    if (this.offlineEnabled()) {
+      // if offline mode is enabled, we notify the frontend
+      //  and proceed as if we are connected
+      return this.invoke_cb('connected', this.offlineMode);
+    }
+    return this.invoke_cb(type, message);
+  }
+
+  // Helper function to invoke the I/O callback.
+  public io_cb(ignored: any, message: any) {
+    return this.invoke_cb('io', message);
+  }
+
+  public test_cb(ignored: any, message: any) {
+    return this.invoke_cb('test', message);
+  }
+
+  public isConnected(): boolean {
+    return this.connected;
+  }
+
+  public isOffline(): boolean {
+    return this.offlineMode === OfflineMode.Forced || (!this.connected && this.offlineMode === OfflineMode.On);
+  }
+
+  public offlineEnabled(): boolean {
+    return this.offlineMode === OfflineMode.On || this.offlineMode === OfflineMode.Forced;
+  }
+
+  public setOfflineModeSetting(setting: number): void {
+    var old = this.offlineMode;
+    if (setting === 0 || setting === 1 || setting === 2) {
+      this.offlineMode = setting;
+      // Set cookie, set expiry date to some date sufficiently in the future.
+      var expiryDate = new Date();
+      expiryDate.setFullYear(expiryDate.getFullYear() + 10);
+      /* TODO set cookie here
+      cookies.put(SEASHELL_OFFLINE_MODE_COOKIE, setting,
+        {secure: true, expires: expiryDate});
+      */
+    } else {
+      throw new WebsocketError("Invalid offline mode setting selected.", setting);
+    }
+    if (old === 2 && this.offlineMode !== old) {
+      // trigger reconnect and sync
+      if (!this.connected)
+        this.connect();
+      else {
+        this.invoke_cb('connected');
+        this.syncAll();
+      }
+    } else if (this.offlineMode === 2) {
+      this.invoke_cb('connected', this.offlineMode);
+    }
+  }
+
+  private make_offline_disabled(name: string) {
+    return async (...args: any[]) => {
+      if (!this.isOffline()) {
+        var res = await this.socket[name].apply(this.socket, args);
+        await this.syncAll();
+        return res;
+      } else {
+        this.rejectOffline(name);
+      }
+    };
+  }
+
+  private make_offline_noop(name: string) {
+    return async (...args: any[]) => {
+      if (!this.isOffline()) {
+        return this.socket[name].apply(this.socket, args);
+      } else {
+        return true;
+      }
+    };
+  }
+
+  private make_offline_enabled(name: string, offlineWriteThrough?: boolean) {
+    const online_arity = this.socket[name].length;
```

Comment:
Yes they do not type check properly. And it loses the ability to extend classes.
I member I also had problem catching errors (can't recall exactly).

Yeah sounds good I will add handlers for the offline logics. This is just a fast merge so that Kevin can work on the frontend.